promql/engine.go

```diff
@@ -1098,44 +1108,65 @@ func (ev *evaluator) aggregation(op itemType, grouping model.LabelNames, without
 					Copied: true,
 				}
 				for _, l := range grouping {
-					if v, ok := sample.Metric.Metric[l]; ok {
+					if v, ok := s.Metric.Metric[l]; ok {
 						m.Set(l, v)
 					}
 				}
 			}
 			result[groupingKey] = &groupedAggregation{
 				labels:           m,
-				value:            sample.Value,
-				valuesSquaredSum: sample.Value * sample.Value,
+				value:            s.Value,
+				valuesSquaredSum: s.Value * s.Value,
 				groupCount:       1,
 			}
+			if op == itemTopK {
```

Comment:
This method/type is blowing up a fair bit now. Time for a refactoring.
If you are not keen to do it in here, could you file an issue?